File: leetcode/ValidBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from typing import Optional

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Determines if a binary tree is a valid Binary Search Tree (BST).

        A valid BST has the following properties:
        1. The left subtree of a node contains only nodes with values less than the node's value.
        2. The right subtree of a node contains only nodes with values greater than the node's value.
        3. Both the left and right subtrees must also be binary search trees.

        This solution uses a recursive approach, passing down the valid range (low, high)
        for each node's value.
        """
        def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
            """
            Recursively validates the BST property for a subtree.
            - node: The current node to check.
            - low: The lower bound (exclusive) for the node's value.
            - high: The upper bound (exclusive) for the node's value.
            """
            # An empty node is a valid part of a BST.
            if not node:
                return True

            # The node's value must be within the valid range defined by its ancestors.
            if not (low < node.val < high):
                return False

            # Recursively check the left and right subtrees with updated bounds.
            # For the left child, the new upper bound is the current node's value.
            # For the right child, the new lower bound is the current node's value.
            return validate(node.left, low, node.val) and validate(node.right, node.val, high)

        # Start the validation from the root with an infinite range.
        return validate(root, -float("inf"), float("inf"))

File: leetcode/medium/ValidBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    /**
     * Checks if a binary tree is a valid Binary Search Tree (BST).
     *
     * @param root The root node of the binary tree.
     * @return true if the tree is a valid BST, false otherwise.
     */
    public boolean isValidBST(TreeNode root) {
        // Start the recursive validation with no bounds (null represents -infinity and +infinity).
        // This correctly handles all possible integer values, including MIN_VALUE and MAX_VALUE.
        return validate(root, null, null);
    }

    /**
     * Helper function to recursively validate the tree within a given range.
     *
     * A node is valid if its value is strictly between the min and max bounds
     * imposed by its ancestors.
     *
     * @param node The current node to validate.
     * @param min The *exclusive* lower bound (node.val must be > min). A null value means no lower bound.
     * @param max The *exclusive* upper bound (node.val must be < max). A null value means no upper bound.
     * @return true if the subtree rooted at 'node' is valid, false otherwise.
     */
    private boolean validate(TreeNode node, Integer min, Integer max) {
        // Base case: An empty (null) node is a valid BST.
        if (node == null) {
            return true;
        }

        // Check 1: Validate against the lower bound.
        // If 'min' is not null, the node's value must be strictly greater than it.
        if (min != null && node.val <= min) {
            return false;
        }

        // Check 2: Validate against the upper bound.
        // If 'max' is not null, the node's value must be strictly less than it.
        if (max != null && node.val >= max) {
            return false;
        }

        // Recursively check the left and right subtrees, tightening the bounds.
        return
            // For the left child, the new *max* bound is this node's value.
            validate(node.left, min, node.val) &&
            // For the right child, the new *min* bound is this node's value.
            validate(node.right, node.val, max);
    }
}